### ISSUE TITLE
Caching ``__dask_graph__``

### ISSUE THREAD

**Issue by phofl** [dask-contrib/dask-expr#183]
Currently, we are building the graph multiple times before doing the actual computation. This wasn't a problem, since the optimizations were cheap. We should cache the graph to avoid multiple passthroughs as soon as we start adding methods that might trigger computations during the optimization step (see #181)

**Comment by rjzamora** [comment]
>we are building the graph multiple times before doing the actual computation.

It probably makes sense to cache the graph, but can you provide a simple example where this is a problem?

**Comment by phofl** [comment]
In compute:

``unpack_collections`` calls ``is_dask_collection``. ``is_dask_collection`` calls ``__dask_graph__`` and discards the result. After this we create the graph again which is used to do the actual computation. This is problematic if ``__dask_graph__`` triggers computations and isn't cheap anymore

**Comment by mrocklin** [comment]
For context, I think that this is only this way because of Xarray.  Previously we had something like the following:

```python
def is_dask_collection(x):
    return hasattr(x, "__dask_graph__")
```

However xarray.DataSet objects can be either dasky or non-dasky depending on some of their values, hence the call. 

I think that the right thing to do here isn't to cache, but instead to change the behavior of `is_dask_collection`. I propose that we special-case xarray, and do something like the above in the common-case.

**Comment by rjzamora** [comment]
Okay, thanks. That's useful context.  I completely agree that we should avoid calling `.dask`/`.__dask_graph__()` in `is_dask_collection` to make this problem much less common.

**Comment by mrocklin** [comment]
@dcherian who is a good xarray person to ping on issues like this these days?  Is it you?  Joe?  Someone else?

**Comment by dcherian** [comment]
Yes, Stephan , Tom Nicholas, and I are most familiar with this stuff, I think.

**Comment by dcherian** [comment]
Should there be a `def __is_dask_collection__(self) -> bool` method instead? We could implement that using a [cheaper check](https://github.com/pydata/xarray/blob/86f99337d803866a4288fc7550f9ee8c495baf87/xarray/core/pycompat.py#L94-L95)

**Comment by mrocklin** [comment]
I'd be fine with that.  cc'ing @fjetter now that this is getting into Dask protocols.

**Comment by fjetter** [comment]
I'm not too deeply involved with dask-expr so far but I would rather suggest changing dask+distributed to ensure there is just a single place where the graphs materialization is triggered than to cache. Caching is simpler but we suffered this kind of accidental materialization problem in the past.

I did a bit of research into what is necessary in dask+distributed to streamline all of this, see https://github.com/dask/distributed/pull/7942
A more or less interesting albeit obvious observation was that concepts right now are a bit misleading. Right now we have a collection, a non-materialized graph, a materialized graph and a low level graph. Particularly the degeneration of graphs is harmful as I'm sure you are all aware of.
Instead, the future should likely only have a collection-like and the low level graph but nothing in between. This is a conversation I'd like to get some input on over there because I can prepare all necessary changes for dask-expr then.


About the matter at hand, I'm fine with introducing `__is_dask_collection__`. I do wonder, though, why we are choosing (or were historically) duck typing over inheritance. Something like https://github.com/dask/distributed/pull/7942#issuecomment-1602703700 could be the minimal base class/protocol others inherit from.

**Comment by dcherian** [comment]
>  be surprised if they're amenable to depending on dask/dask at this point.

Don't think Xarray will. If anything, we're moving towards removing the pandas dependency.


> bout the matter at hand, I'm fine with introducing __is_dask_collection__ but I do wonder why we are choosing (or were historically) duck typing over inheritance. 

Just to be clear, we'll still need something like `__is_dask_collection__` for Xarray in either case. A plain `isinstance` check would not work. For example:

```python
import dask
import xarray as xr

ds = xr.tutorial.open_dataset("air_temperature")  # needs pooch
dask.base.is_dask_collection(ds)  # False

ds["air"] = ds.air.chunk(time=24)
dask.base.is_dask_collection(ds)  # True
```
